### PR TITLE
[Main Map] Only display job cards if below certain zoom level

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Towards #1348 

@scolsen  @gino-m  PTAL?

### CODE REVIEW

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerViewModel.kt [hunk 1]
```diff
@@ -194,23 +150,15 @@ internal constructor(
...
-
-    val savedMapLocationsOfInterest =
-      Flowable.combineLatest(
-        loiStream.map { locationsOfInterest -> toLocationOfInterestFeatures(locationsOfInterest) },
-        selectedLocationOfInterest
-      ) { locationsOfInterest, selectedLocationOfInterest ->
-        updateSelectedLocationOfInterest(locationsOfInterest, selectedLocationOfInterest)
-      }
 
+    // TODO: Should we only render the LOIs that are present within the map bounds?
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerViewModel.kt:154]: This is a bigger project imo and not actionable here. Can you move this to an issue instead of a TODO?

File: ground/src/main/java/com/google/android/ground/ui/map/LoiController.kt [hunk 2]
```diff
@@ -30,27 +28,23 @@ import io.reactivex.Flowable
...
 internal constructor(
   private val surveyRepository: SurveyRepository,
   private val locationOfInterestRepository: LocationOfInterestRepository,
 ) {
 
   private val cameraBoundsSubject: @Hot Subject<LatLngBounds> = PublishSubject.create()
-  val locationsOfInterest: LiveData<List<LocationOfInterest>>
 
-  init {
+  fun loisWithinMapBounds(): Flowable<List<LocationOfInterest>> {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/map/LoiController.kt:43]: Can this be moved into the loiRep as `getWithinBounds(LatLngBounds)`? The camera bounds should probably live in the `MapStateRepository` or `BaseMapViewModel`, then this class can be deleted.

**Review by gino-m — CHANGES_REQUESTED** [review]

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt [hunk 3]
```diff
@@ -127,4 +124,33 @@ constructor(
...
+    surveyRepository.activeSurvey
+      .switchMap { survey ->
+        survey
+          .map { localLoiStore.getLocationsOfInterestOnceAndStream(it) }
+          .orElse(Flowable.just(setOf()))
+      }
+      .distinctUntilChanged()
+
+  /** Returns a flowable of all [LocationOfInterest] within the map bounds (viewport). */
+  fun getWithinBoundsOnceAndStream(
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:139]: Camera bounds sounds like something that belongs in the view or in MapStateRepository. Can we pass that in as an arg here instead of maintaining it in this class?

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt [hunk 4]
```diff
@@ -127,4 +124,33 @@ constructor(
...
+    cameraBoundUpdates: Flowable<LatLngBounds>
+  ): Flowable<List<LocationOfInterest>> {
+    val loiStream = getAllLocationsOfInterestOnceAndStream()
+    return cameraBoundUpdates
+      .flatMap { bounds -> loiStream.map { it.toLoiCardsWithinBounds(bounds) } }
+      .distinctUntilChanged()
+  }
+
+  /** Filters all [LocationOfInterest] within [bounds]. */
+  private fun Set<LocationOfInterest>.toLoiCardsWithinBounds(
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:149]: Cards are also a UI element that is out of scope of the repository layer. Maybe it's just a naming issue, and this should be `getLoisWithinBounds()`?

File: ground/src/main/java/com/google/android/ground/ui/common/BaseMapViewModel.kt [hunk 5]
```diff
@@ -48,6 +55,13 @@ constructor(
   mapController: MapController
 ) : AbstractViewModel() {
 
+  private val cameraZoomSubject: @Hot Subject<Float> = PublishSubject.create()
+  val cameraZoomUpdates: Flowable<Float> = cameraZoomSubject.toFlowable(BackpressureStrategy.LATEST)
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/common/BaseMapViewModel.kt:59]: Consider using `SharedFlow` for new event streams. Feel free to change in a follow-up if that causes too much churn in this PR.

**Review by gino-m — CHANGES_REQUESTED** [review]
Great! Thanks for the updates.

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt (hunk 3 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:139]: `LatLngBounds` flowable is already passed via an arg. Are you referring to something else?

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt (hunk 4 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt]: Removed mentions of cards from repository. Thanks!

File: ground/src/main/java/com/google/android/ground/ui/common/BaseMapViewModel.kt (hunk 5 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/ui/common/BaseMapViewModel.kt:59]: I'll file a bug to do it separately

**Comment by shobhitagarwal1612** [comment]
@gino-m Gentle ping!

**Review by scolsen — APPROVED** [review]

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt (hunk 3 above)
Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:137]: Sorry, I missed that the bounds were being passed in as an arg. Please disregard.

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt [hunk 6]
```diff
@@ -127,4 +124,33 @@ constructor(
       MutationEntitySyncStatus.IN_PROGRESS,
       MutationEntitySyncStatus.FAILED
     )
+
+  /** Returns a flowable of all [LocationOfInterest] for the currently active [Survey]. */
+  fun getAllLocationsOfInterestOnceAndStream(): Flowable<Set<LocationOfInterest>> =
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:127]: Can we update this to only look at the offline LOIs? I'm working on code that will update those in the background, so we no longer need to subscribe to updates from Firestore here; instead we can subscribe to changes in the local Room store.

**Review by gino-m — CHANGES_REQUESTED** [review]

File: ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt (hunk 6 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:127]: This does return only offline LOIs. Should I rename it to something else to make it more explicit?

Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/repository/LocationOfInterestRepository.kt:125]: It's ok for now, I'll update it in my upcoming PR. Thanks!

**Review by gino-m — APPROVED** [review]
🔍 ↕️
